unixcreds.go

```diff
@@ -0,0 +1,90 @@
+// +build linux freebsd solaris
+
+package ttrpc
+
+import (
+	"context"
+	"net"
+	"os/user"
+	"strconv"
+
+	"github.com/pkg/errors"
+	"golang.org/x/sys/unix"
+)
+
+var (
+	UnixSocketRequireSameUser = UnixCredentialsFunc(requireSameUser)
+	UnixSocketRequireRoot     = UnixCredentialsFunc(requireRoot)
+)
+
+type UnixCredentialsFunc func(*unix.Ucred) error
+
+func (fn UnixCredentialsFunc) Handshake(ctx context.Context, conn net.Conn) (net.Conn, interface{}, error) {
+	uc, err := requireUnixSocket(conn)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	fp, err := uc.File()
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "unixSocketHandshake: failed to get unix file")
+	}
+	defer fp.Close() // this gets duped and must be closed when this method is complete.
+
+	ucred, err := unix.GetsockoptUcred(int(fp.Fd()), unix.SOL_SOCKET, unix.SO_PEERCRED)
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "unixSocketHandshake: failed to retrieve socket peer credentials")
+	}
+
+	if err := fn(ucred); err != nil {
+		return nil, nil, err
+	}
+
+	return uc, ucred, nil
+}
+
+func UnixSocketRequireUidGid(uid, gid uint32) UnixCredentialsFunc {
+	return func(ucred *unix.Ucred) error {
+		return requireUidGid(ucred, uid, gid)
+	}
+}
+
+func requireRoot(ucred *unix.Ucred) error {
+	return requireUidGid(ucred, 0, 0)
+}
+
+func requireSameUser(ucred *unix.Ucred) error {
+	u, err := user.Current()
+	if err != nil {
+		return errors.Wrapf(err, "could not resolve current user")
```

Comment:
prepend `unixSocketHandshake:` to these errors?

Reply:
I'll make sure they are wrapped in the next level.